### ISSUE TITLE
Question: Constant Prediction Value in ExactGP

### ISSUE THREAD

**Issue by GStechschulte** [cornellius-gp/gpytorch#1906]
Hello, I am attempting to use ExactGP for time series forecasting in which I have the following synthetic seasonal data:

```python
# Number of samples. 
n = 1000
# Generate "time" variable. 
t = np.arange(n)
data_df = pd.DataFrame({'t' : t})

# Generate seasonal variables. 
def seasonal(t, amplitude, period):
    """Generate a sinusoidal curve."""
    y1 = amplitude * np.sin((2*np.pi)*t/period) 
    return y1

# Add two seasonal components. 
data_df['s1'] = data_df['t'].apply(lambda t : seasonal(t, amplitude=2, period=40))

# Define target variable. 
data_df['y1'] = data_df['s1']

# Set noise standard deviation. 
sigma_n = 0.3

data_df['epsilon'] = np.random.normal(loc=0, scale=sigma_n, size=n)
# Add noise to target variable. 
data_df['y1'] = data_df['y1'] + data_df ['epsilon']
```

![output](https://user-images.githubusercontent.com/63432018/152502602-75376e77-be20-4dd0-8f7d-58b0592955f3.png)

Then, converting the data to train and test:

```python
X = data_df['t'].values
y = data_df['y1'].values

prop_train = 0.7
n_train = round(prop_train * n)

X_train = torch.from_numpy(X[:n_train]).to(torch.float32)
y_train = torch.from_numpy(y[:n_train]).to(torch.float32)

X_test = torch.from_numpy(X[n_train:]).to(torch.float32)
y_test = torch.from_numpy(y[n_train:]).to(torch.float32)
```

Copying along with the ExactGP tutorial and as the data is synthetically generated with a seasonal component, I have decided to use the `gpytorch.kernels.PeriodicKernel()` with the following priors on the length scale and period (the priors are chosen based on the fact that I know how the data was generated, i.e., period=40):

```python
# We will use the simplest form of GP model, exact inference
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):

        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)

        self.mean_module = gpytorch.means.ConstantMean()
        lengthscale_prior = gpytorch.priors.GammaPrior(0.5, 1.5)
        period_prior = gpytorch.priors.NormalPrior(40, 1)
        period_constraint = [35, 45]

        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.PeriodicKernel(
                lengthscale_prior=lengthscale_prior,
                period_prior=period_prior,
                period_constraint=period_constraint
                )
            )
    
    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(X_train, y_train, likelihood)
```

Using the same training loop as in the ExactGP tutorial:

```python
# this is for running the notebook in our testing framework
import os
smoke_test = ('CI' in os.environ)
training_iter = 2 if smoke_test else 50

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)  # Includes GaussianLikelihood parameters

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

for i in range(training_iter):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model(X_train)
    # Calc loss and backprop gradients
    loss = -mll(output, y_train)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (
        i + 1, training_iter, loss.item(),
        model.covar_module.base_kernel.lengthscale.item(),
        model.likelihood.noise.item()
    ))
    optimizer.step()
```
where the following is the loss and raw_lengthscale and raw_noise learned from training:
```
Iter 1[/50]() - Loss: 2.251   lengthscale: 0.693   noise: 0.693
Iter 2[/50]() - Loss: 2.184   lengthscale: 0.744   noise: 0.744
Iter 3[/50]() - Loss: 2.124   lengthscale: 0.798   noise: 0.798
Iter 4[/50]() - Loss: 2.072   lengthscale: 0.854   noise: 0.853
Iter 5[/50]() - Loss: 2.028   lengthscale: 0.911   noise: 0.910
Iter 6[/50]() - Loss: 1.985   lengthscale: 0.968   noise: 0.968
Iter 7[/50]() - Loss: 1.957   lengthscale: 1.026   noise: 1.027
Iter 8[/50]() - Loss: 1.929   lengthscale: 1.082   noise: 1.087
Iter 9[/50]() - Loss: 1.905   lengthscale: 1.136   noise: 1.147
Iter 10[/50]() - Loss: 1.886   lengthscale: 1.186   noise: 1.206
Iter 11[/50]() - Loss: 1.869   lengthscale: 1.233   noise: 1.265
Iter 12[/50]() - Loss: 1.856   lengthscale: 1.277   noise: 1.323
Iter 13[/50]() - Loss: 1.844   lengthscale: 1.317   noise: 1.380
Iter 14[/50]() - Loss: 1.834   lengthscale: 1.352   noise: 1.435
Iter 15[/50]() - Loss: 1.826   lengthscale: 1.382   noise: 1.489
Iter 16[/50]() - Loss: 1.819   lengthscale: 1.406   noise: 1.541
Iter 17[/50]() - Loss: 1.814   lengthscale: 1.425   noise: 1.591
Iter 18[/50]() - Loss: 1.809   lengthscale: 1.438   noise: 1.639
Iter 19[/50]() - Loss: 1.806   lengthscale: 1.446   noise: 1.685
Iter 20[/50]() - Loss: 1.802   lengthscale: 1.448   noise: 1.729
Iter 21[/50]() - Loss: 1.800   lengthscale: 1.444   noise: 1.770
Iter 22[/50]() - Loss: 1.798   lengthscale: 1.435   noise: 1.809
Iter 23[/50]() - Loss: 1.796   lengthscale: 1.421   noise: 1.846
Iter 24[/50]() - Loss: 1.795   lengthscale: 1.402   noise: 1.881
Iter 25[/50]() - Loss: 1.794   lengthscale: 1.379   noise: 1.914
Iter 46[/50]() - Loss: 1.787   lengthscale: 0.529   noise: 2.209
Iter 47[/50]() - Loss: 1.787   lengthscale: 0.494   noise: 2.211
Iter 48[/50]() - Loss: 1.787   lengthscale: 0.461   noise: 2.212
Iter 49[/50]() - Loss: 1.787   lengthscale: 0.430   noise: 2.212
Iter 50[/50]() - Loss: 1.786   lengthscale: 0.400   noise: 2.212
```

and the same prediction code:

```python
# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Make predictions by feeding model through likelihood
with torch.no_grad(), gpytorch.settings.fast_pred_var():
    observed_pred = likelihood(model(X_test))
```

However, the prediction is just a constant value, which I don't understand as I have used the periodic kernel with a prior length scale and periodic component. I have also varied the learning rate from 0.1 to 0.01, 0.001 and also ran 100 iterations; still resulting in the same constant value being predicted:

```python
with torch.no_grad():
    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(16, 6))
    # Plot training data as black stars
    ax.plot(X_train.numpy(), y_train.numpy())
    # Plot predictive means as blue line
    ax.plot(X_test.numpy(), observed_pred.mean.numpy(), 'b')
    ax.set_ylim([-3, 3])
    ax.legend(['Observed Data', 'Mean', 'Confidence'])
```

![output](https://user-images.githubusercontent.com/63432018/152503886-3eadd6a1-dbd7-45f7-9375-b8dd63a32938.png)

And when looking closely, the constant predictions does show some very very "small" periodicity, likely resulting from the small length scale learned in training (after 50 iterations, the lengthscale=0.40 and noise=2.21). The constant values from prediction make me think it may have something to do with my processing of train and test data?

Thanks for the help!

**Comment by wjmaddox** [comment]
Hmm... this is because the initial period length is ~0.4. setting it to be 40 on initialization resolves the issue:

```python
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.PeriodicKernel(
                lengthscale_prior=lengthscale_prior,
                period_prior=period_prior,
                period_constraint=period_constraint
                )
            )
        self.covar_module.base_kernel.period_length = 40.
```
in the model definition code

Training then gives:
![image](https://user-images.githubusercontent.com/13008542/152650857-e09024be-293a-4d89-9f37-c15d65545f3c.png)

I am not entirely sure why the constraint didn't reset the period length to be something reasonable.

**Comment by wjmaddox** [comment]
> I am not entirely sure why the constraint didn't reset the period length to be something reasonable.

This is because you didn't properly set the constraint which should be set in the following way:
```python
        period_constraint = gpytorch.constraints.Interval(35, 45)

        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.PeriodicKernel(
                lengthscale_prior=lengthscale_prior,
                period_prior=period_prior,
                period_length_constraint=period_constraint
                )
            )
```

**Comment by GStechschulte** [comment]
> > I am not entirely sure why the constraint didn't reset the period length to be something reasonable.
> 
> This is because you didn't properly set the constraint which should be set in the following way:
> 
> ```python
>         period_constraint = gpytorch.constraints.Interval(35, 45)
> 
>         self.covar_module = gpytorch.kernels.ScaleKernel(
>             gpytorch.kernels.PeriodicKernel(
>                 lengthscale_prior=lengthscale_prior,
>                 period_prior=period_prior,
>                 period_length_constraint=period_constraint
>                 )
>             )
> ```
> 
> which removes the need to forcibly set the period length on initialization.


Ah, this makes sense. I just reproduced your solution and now it works for me—thanks for the help! Closing this "issue" now.
